pkg/asset/machines/openstack/machines.go

```diff
@@ -107,7 +108,69 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
 		machines = append(machines, machine)
 	}
 
-	return machines, nil
+	controlPlaneMachineSet := &machinev1.ControlPlaneMachineSet{}
+	replicas := int32(total)
+	failureDomains := []machinev1.OpenStackFailureDomain{}
+
+	if len(mpool.FailureDomains) > 0 {
+		for _, fd := range mpool.FailureDomains {
+			domain := machinev1.OpenStackFailureDomain{}
+			if fd.ComputeAvailabilityZone != "" {
+				domain.ComputeAvailabilityZone = fd.ComputeAvailabilityZone
+			}
+			if fd.StorageAvailabilityZone != "" {
+				domain.StorageAvailabilityZone = fd.StorageAvailabilityZone
+			}
+			failureDomains = append(failureDomains, domain)
+		}
+
+		controlPlaneMachineSet = &machinev1.ControlPlaneMachineSet{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "machine.openshift.io/v1",
+				Kind:       "ControlPlaneMachineSet",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-machine-api",
+				Name:      "cluster",
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-cluster": clusterID,
+				},
+			},
+			Spec: machinev1.ControlPlaneMachineSetSpec{
+				Replicas: &replicas,
+				Selector: metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"machine.openshift.io/cluster-api-cluster":      clusterID,
+						"machine.openshift.io/cluster-api-machine-role": role,
+						"machine.openshift.io/cluster-api-machine-type": role,
+					},
+				},
+				Template: machinev1.ControlPlaneMachineSetTemplate{
```

Comment:
We still are in the `if len(mpool.FailureDomains) > 0` block here. Shouldn't we add the machine template to CPMS regardless of the failureDomains?